compute_yearly_risk.py

```diff
@@ -0,0 +1,396 @@
+"""Script for computing yearly risk for a given model, scenario, and years."""
+
+import argparse
+import time
+from itertools import product
+from multiprocessing import Pool
+from pathlib import Path
+import numpy as np
+import xarray as xr
+
+
+def univoltine(tmin, tmax):
+    """The function used for calculating the summer risk component, the 'survival' associated with univoltinism.
+    
+    Args:
+        tmin (numpy.ndarray): a 1-D array of minimum temperatures for a given year
+        tmax (numpy.ndarray): a 1-D array of maximum temperatures for a given year
+        
+    Returns:
+        the summer risk component as a float between 0 and 1
+    """
+    try:
+        idx = np.where(tmax >= 16)[0][0]
+    except IndexError:
+        if np.isnan(tmin[0]):
+            return np.nan
+        else:
+            return 0
+
+    tmin = tmin[idx + 40 : idx + 90]
+    tmax = tmax[idx + 40 : idx + 90]
+    # hour counter
+    k = 0
+    
+    # easy if tmin is ever above 17 - count entire day
+    # need to remember indices of values so we can exclude from
+    #  the hourly estimator below
+    hot_idx = tmin > 17
+    k += 24 * hot_idx.sum()
+    # discard indices that counted for entire days above 17C
+    tmax = tmax[~hot_idx]
+    tmin = tmin[~hot_idx]
+    
+    # also can immediately discard days where tmax is > 17C
+    cold_idx = tmax < 17
+    # discard indices that counted for entire days above 17C
+    tmax = tmax[~cold_idx]
+    tmin = tmin[~cold_idx]
+    
+    # need special treatment for tmin == 17 as well, as it would
+    #  require division by zero in our estimation algorithm below.
+    #  just assume it is above 17 for 75% of the time, or 18 hrs
+    equal_idx = tmin == 17
+    k += 18 * equal_idx.sum()
+    # discard indices that counted for days where tmin == 17C
+    tmax = tmax[~equal_idx]
+    tmin = tmin[~equal_idx]
+    
+    # finally, can implement our estimation for days where tmin <17 and tmax>17
+    h_est = (24 * (tmax - 17) / ((tmax - 17) + (17 - tmin)))
+    h_est[h_est < 0] = 0
+    k += h_est.sum()
+
+    # then determine "survival" due to univoltinism
+    if k < 40:
+        u = 0
+    elif 40 <= k < 225:
+        u = (k - 40) / 7.4
+    elif 225 <= k < 412:
+        u = 25 + (k - 225) / 2.5
```

Comment:
Really great point / question, thank you! Since these functions are so tightly interlinked with these magic numbers, I opted to leave them in the functions, as I sort of view this file itself as a sort of config file for the risk computations. But I did go through and add comments and such the hopefully makes those numbers a bit more clear. Going to make a ticket to update this repo with references to the "science" when it is available.